### ISSUE TITLE
Source-level moddability discussion

### ISSUE THREAD

**Issue by heinrich5991** [ddnet/ddnet#7777]
How much do we want to support moddability in the DDNet codebase? Note that this doesn't talk about the network protocol but about the actual code.

**Comment by heinrich5991** [comment]
I don't feel like adding interfaces purely to support modding. #7770 is such a PR. The moddability is present on the network level but not inside the code base.

A way to script DDNet would be fine by me, but not piecemeal adding hooks to the codebase.

**Comment by ChillerDragon** [comment]
What should be in and out of scope for ddnet codebase moddability. What should ddnet offer for custom forks? And what will ddnet not do for custom forks?

I personally would like to see most standard mods: ctf/fng/city/ball
With features like: point score, round based, team based, custom tiles, custom database, custom player/character methods and attributes, custom chat/rcon commands, vanilla weapons (damaga/gun curvature/shotgun bullets)

And all of that would be ideally addable to the ddnet codebase without creating any git conflict surface. In for example a own gamecontroller.

Another option would be extending the anti bot code to load more than one antibot module. And rename it from antibot to plugin.

I am currently trying to develop a git conflict driven [ddnet instagib ](https://github.com/ZillyInsta/ddnet-insta/)mod based on ddnet. Trying to touch as little ddnet code as possible to have a clean and updatable ddnet mod. After already having opend a few prs to bend ddnet into allowing mods to do such things its time to set some goal to reduce later discussion times.

The teeworlds project is currently on ice. I tried to contact Oy multiple times but he refuses to join irc or merge some prs. The prs are piling. The code is starting to smell really bad. Everything is deprecated even the https://github.com/teeworlds/teeworlds/pull/3215. The CI is failing and the devs are leaving. It is so bad that I thought of creating a community fork where all of the prs could be merged by active volunteers. I pitched this to @heinrich5991 and he said there is already a community fork: ddnet. Sure that also works but then ddnet has to achieve at least the same support for dm/ctf based mods as teeworlds.

**Comment by ChillerDragon** [comment]
>  allow the mod to change the whole snapping.

Might sound nice. But if the whole snapping is a lot of code the mod has to reimplement/copy/maintain all of that code when updating the mod. Imo a few standard things should be maintained by ddnet so that all basic mods can get free ddnet updates.

**Comment by ChillerDragon** [comment]
> > Might sound nice. But if the whole snapping is a lot of code the mod has to reimplement/copy/maintain all of that code when updating the mod.
> 
> They don't have to update that code, they can leave it as-is. We have backcompat on the client side for these cases.

There are also compile errors and git conflicts that I am worried about. And then parts of the mod are outdated which is not nice (still better than no update at all).

**Comment by ChillerDragon** [comment]
The thing is if I need to overwrite 20 lines of snap just to add point based scores again. I know that this mod needs attention when merged with upstream a few years later. If one cares about updating to upstream the teeworlds project becomes a more interesting code base. If ddnet wants to keep the "we are an own game" "we stay in 0.6" "nobody plays 0.7 anyways" agenda. it would ideally include that all mods such as zCatch, fng, iCTF, teeball, City, Block, Infection, .. are based on ddnet. And that should be officially supported by ddnet including having the option to update ddnet forks without spending 1 hour on every merge.

**Comment by Kaffeine** [comment]
As a mod maintainer (not author) I see a lot of resources wasted on syncing the core game codebase. It would be amazing for the mods, if if DDNet would be a "base game with mod API **+** DDRace gametype implemented using that API". We all could port our mods to that API and sync the base game with a magnitude (or two) less efforts.

I see that it is much easier to mix the gameplay-specific code into the base classes and call it a day. I see that the current game architecture does not allow to reasonably split the code (responsibilities), e.g. GameContext has a lot of mod-specific code and IGameController has no API to override it. Also, as soon as you change CCharacter, CPlayer, and other base classes (compared to vanilla), you should realise that a mod is probably also need to change those classes/those parts. Now you also see that without mod API virtually all mods 'll have tons of git conflicts whenever you adjust ddrace-specifc code.

I see a mod API as C++ interfaces, which are provided by the base game and used by all gameplay implementations. For the first step which hugely improve the situation, we (or I, as a mod dev) want to have an API to override the base classes behavior (with ddrace gameplay implemented in those classes). For the next step we can think about moving ddrace stuff to subclasses, which means components decoupling, profitable for DDNet.

@heinrich5991 
> Instead of making an interface to change the client_score_kind to a value, add an interface to add arbitrary items to the game info JSON. Instead of making an interface to change a single flag of the game info object, allow the mod to change the whole snapping.

I support this idea, and this is what https://github.com/ddnet/ddnet/pull/7631 is about. It allows mods to override 'whole things' without hundreds of hooks.

I see how those highly granulated hooks would simplify small mods but IMO it overcomplicates the codebase, and the possible benefits do not worth it. See my comment there: https://github.com/ddnet/ddnet/pull/7534#issuecomment-1825179756 (unfortunately DDNet devs didn't reply).

**Comment by heinrich5991** [comment]
Responding to https://github.com/ddnet/ddnet/pull/7534#issuecomment-1825179756:

> I think that before doing such changes (which solve a particular problem in a utilitarian way) we need to come to a _concept_ on how we see the mods and modding API.

You realized this problem before I did. :)

> We need to define the roles of the GameContext and GameController. Then we can see if the GameContext actually need to behave differently depending on the gametype, and then decide whether it makes sense to delegate that stuff to GameController or we still want to have the roles separated and need another way to deal with that.
> 
> In other words, this complicates the code and even more couples GameController with GameContext but it won't solve the need of other tens customizations which a medium-sized mod still needs to have in GameContext. And if all those issues 'll be solved like that, we'd end up with a tough spaghetti.
> 
> We barely can use this approach to allow a mod to add new entities, map layers, chat/rcon commands, etc and _maybe_ the recommended way would be class inheritance. Instead of wiring up the net messages to GameController, we can say that a medium-large scale mod which needs to customize this, should subclass CGameContext and override the needed methods right there. The classes roles 'll be kept separated. The 'maintainance burden' and modding support in DDNet 'll be limited by `virtual` keyword added here and there.
> 
> Closer to the code: on one hand, it is indeed allowing a controller to intercept those messages. On the other hand, it leads to API duplication; maybe from the code point of view it makes sense to add `IGameController::OnMessage` and call it from GameContext after TeeHistorian (so the controller can also hook `OnStartInfoNetMessage()`).

That sounds pretty sane to me.

**Comment by ChillerDragon** [comment]
I am cool with subclassing CGameContext. Would be more comfortable with it if I would be able to play with the actual code for it. Will the official ddnet repo have a placeholder example such as mod.cpp? How much code has to be touched that might git conflict when subclassing CGameContext? Are prs in scope of ddnet then that improve subclassed mod maintenance. Such as minimizing the risk for build and logic errors by splitting ddnet code in small methods that can be overwritten. So there is less duplicated code that has to be maintained by mod developers.

**Comment by heinrich5991** [comment]
> Will the official ddnet repo have a placeholder example such as mod.cpp?

Not sure, tending towards "no".

> Are prs in scope of ddnet then that improve subclassed mod maintenance. Such as minimizing the risk for build and logic errors by splitting ddnet code in small methods that can be overwritten. So there is less duplicated code that has to be maintained by mod developers.

Also tending towards "no".

**Comment by Kaffeine** [comment]
> Are prs in scope of ddnet then that improve subclassed mod maintenance. Such as minimizing the risk for build and logic errors by splitting ddnet code in small methods that can be overwritten. So there is less duplicated code that has to be maintained by mod developers.

I think that most of such changes would be beneficial for DDNet as it mostly means normal maintainance/clean — better split onto functions, improved incapsulation, etc. I mean PRs like those: https://github.com/ddnet/ddnet/pull/7147/files, https://github.com/ddnet/ddnet/pull/7136.

IMO it would be enough to limit the explicit mod support in `virtual` keywords added in a few key places, and `private` replaced by `protected` in some base classes.

**Comment by Kaffeine** [comment]
>> Will the official ddnet repo have a placeholder example such as mod.cpp?
>
> Not sure, tending towards "no".

It would be better for the mods community if we'll eventually move DDNet stuff to CDDRaceGameContext (assuming that we accept the subclassing approach). We won't need an example of mod gamecontext if DDNet would use the same mechanism (like we already have `CGameControllerDDRace`). For DDNet it means better design with better responsibility separation (opposed to the current monoliths).

**Comment by heinrich5991** [comment]
> It would be better for the mods community if we'll eventually move DDNet stuff to CDDRaceGameContext (assuming that we accept the subclassing approach).

Can you give an example of such behavior?

**Comment by Kaffeine** [comment]
> > It would be better for the mods community if we'll eventually move DDNet stuff to CDDRaceGameContext (assuming that we accept the subclassing approach).
> 
> Can you give an example of such behavior?

Can you please clarify the question? Do you mean "example of ddrace-specific behavior in GameContext"?

### Commands

The first thing comes to mind is commands registration. Race-specifc commands can be registered in CDDRaceGameContext OnConsoleInit() override. Some commands in `ddracecommands` / `ddracechat` are general purpose, and some commands inlined in `CGameContext::OnConsoleInit()` (namely `random_unfinished_map`) are totally mod-specific.

### NetMessages

E.g. `CGameContext::OnKillNetMessage` with `pPlayer->Respawn();` does too much for mods with spawn cooldown (including vanilla gametypes).

### Tunezones

Tunezones are ddrace specific / different from vanilla. Whole "Tuning Zones" thing is a total mess with code and variables duplications and bugs.
```cpp
 	// Reset Tunezones
	CTuningParams TuningParams;
	for(int i = 0; i < NUM_TUNEZONES; i++)
	{
		TuningList()[i] = TuningParams;
		TuningList()[i].Set("gun_curvature", 0);
		TuningList()[i].Set("gun_speed", 1400);
		TuningList()[i].Set("shotgun_curvature", 0);
		TuningList()[i].Set("shotgun_speed", 500);
		TuningList()[i].Set("shotgun_speeddiff", 0);
	}
	// Reset Tuning
	if(g_Config.m_SvTuneReset)
	{
		ResetTuning();
	}
	else
	{
		Tuning()->Set("gun_speed", 1400);
		Tuning()->Set("gun_curvature", 0);
		Tuning()->Set("shotgun_speed", 500);
		Tuning()->Set("shotgun_speeddiff", 0);
		Tuning()->Set("shotgun_curvature", 0);
	}
```
This can be extracted to virtual `InitTuning()` with DDNet implementation kept the same/isolated from the base game context.

### Entities

`CGameContext::CreateAllEntities(bool Initial)` is ddnet specifc. It won't be a big problem for mods to keep this code though.

### Score

Score implementation bits in GameContext are race-specific. This race stuff in generic GameContext blocked score incapsulation (`#3` in  https://github.com/ddnet/ddnet/pull/7598#issuecomment-1846757799).

**Comment by ChillerDragon** [comment]
The tuning thing would be so useful. Creating a simple CTF like mod based on ddnet is such a mess to get the correct gun curvature back.

**Comment by heinrich5991** [comment]
Thanks for the explanations, the examples make a lot of sense.

**Comment by Bamcane** [comment]
If we use DDNet to create a mod now, it's still not easy to clean code that needn't. Also, DDNet-based mod usually build slowly, not be good for mod creater to test their mod many times with file change. And DDNet need to recode many functions to create a PVP mod

**Comment by jxsl13** [comment]
bump

**Comment by BlaiZephyr** [comment]
bump²

**Comment by Kaffeine** [comment]
Understandable and expected. Good luck.

**Comment by BlaiZephyr** [comment]
second this, DDNet shouldn't be cluttered with modding-related PR's with no addition to DDNet itself. Can we expect PR's with the intent to increase moddability (in terms of refactoring or restructuring existing code) to be accepted then?

**Comment by Kaffeine** [comment]
> Can we expect PR's with the intent to increase moddability (in terms of refactoring or restructuring existing code) to be accepted then?

From the subject MR's through comments like [this](https://github.com/ddnet/ddnet/issues/7777#issuecomment-1905866310) as examples of what could be refactored or restructured, I read the answer "decline" as "no, we don't want it".

Extra abstractions have no gameplay value for DDNet itself. Maybe external contributors could cleanup duplications and reduce the dependencies but it comes at cost of maintainers time, and sometimes it would bloat the code. Most of the time it is easier to directly read the functions and values than go through abstractions. It won't be rational to add or keep unused abstractions.

**Comment by hardliner66** [comment]
A huge part why DDNet even exists is because the original was opensource and quite modable. Yes, they didn't care about mods and there was no official support for modding, but at least it was there. In hindsight it would have been better to keep DDNet functionality separate and update/improve the core as needed, to make DDNet work. This could have been a great base for all kinds of mods, but with more modern features than the old 0.6 base.

So while I don't think modding should be the highest priority, I think cleaning up the core and havomg

**Comment by heinrich5991** [comment]
> I think cleaning up the core

That's still appreciated.

> having a well-defined API that still allows for gamemodes that are more similar to the older tw

That's out of scope now (unless DDNet needs it). It was also out of scope for Teeworlds, btw.

**Comment by jxsl13** [comment]
https://github.com/teeworlds/teeworlds/blob/master/src/game/server/gamemodes/mod.cpp

*mic drop*

**Comment by hardliner66** [comment]
> > I think cleaning up the core
> 
> 
> 
> That's still appreciated.
> 
> 
> 
> > having a well-defined API that still allows for gamemodes that are more similar to the older tw
> 
> 
> 
> That's out of scope now (unless DDNet needs it). It was also out of scope for Teeworlds, btw.

It's out of scope because DDNets implementation details already have leaked into the core (last time I checked there were plenty of casts to DDNet classes) and because it's cumbersome to put in the work to define a proper API.

I probably would have done the same, but it's still a shame that the modability suffers from that. Having a stable core with everything else just tacked on would be way better for the community as a whole, as it would make the life of everyone outside of DDNet easier leading to more more diversity which is beneficial for the game.

Maybe we can find a middle ground where we can define modability as some kind of soft goal, where it's fine if someone wants to add improvements, even if they aren't directly beneficial to DDNet, as long as they don't make it worse. This someone could start untangling the core from DDNet while everything else stays intact and with minimal additional work for the staff, as most of those things should be refactorings anyway.

**Comment by hardliner66** [comment]
> https://github.com/teeworlds/teeworlds/blob/master/src/game/server/gamemodes/mod.cpp
> 
> *mic drop* 
> 
> Teeworlds, in the long run, was a better base for mods, because it did not have 5 PRs merged daily.
> 
> Maybe modding died with Teeworlds.

As long as the APIs don't change, most of those changes should be irrelevant. And with a proper deprecation cycle mod devs should have enough time to react.

**Comment by ChillerDragon** [comment]
> last time I checked there were plenty of casts to DDNet classes


I fixed that :3

**Comment by Learath2** [comment]
> Teeworlds has [gamemodes](https://github.com/teeworlds/teeworlds/blob/c56fa9e6a20cfc9d7d16502e18c7d7633acdf492/src/game/server/gamemodes/mod.cpp). I would call that a well-defined API. DDNet has [gamemodes](https://github.com/ddnet/ddnet/blob/f9df4a85be4ca94ca91057cd447707bcce16fd94/src/game/server/gamemodes/mod.cpp) too. But they do not support something "more similar to the older tw". Because adding back vanilla gameplay has to be done by the gamemode author.

tbf you'd be hard pressed to find anyone that actually built any significant mod utilizing the given abstractions, because as you and many others noticed the gamecontroller is just not a "strong" enough component. Basically any mod that has ever added a tile leaks out of the prescribed way to mod the game.

When ddnet7 was actually being considered I did a fair bit of work to separate the gamecontext and the gamecontroller more and have the gamecontext defer to the controller more. However, soon we decided the sixup compatibility layer was a better way to go about supporting 0.7, so I never really looked into it more.

**Comment by ChillerDragon** [comment]
> tbf you'd be hard pressed to find anyone that actually built any significant mod utilizing the given abstractions, because as you and many others noticed the gamecontroller is just not a "strong" enough component. Basically any mod that has ever added a tile leaks out of the prescribed way to mod the game.

[ddnet-insta](https://github.com/ddnet-insta/ddnet-insta) is mostly implemented in the gamecontroller. To be fair there are [more controller hooks](https://github.com/ddnet-insta/ddnet-insta/blob/07cec6018d74f485e44c5ef77448efade81c1eea/src/game/server/instagib/gamecontroller.h) added. But even with pure ddnet you can get quite far. And tiles can already be fully implemented within a gamemode without touching any other file thanks to a14607f818d19a36f14fb5e4f6a15ce7c95765a1. Even zCatch can force keep players spectators with a custom message using `CanJoinTeam()`  https://github.com/ddnet/ddnet/blob/9dbf46c5731f921069a7761baa453ee2779455f8/src/game/server/gamecontext.cpp#L2543

But yes there are a few pain points still. For example rcon/chat commands and configs. And of course the more complex things like custom entities in the world or things deeper in the network/server stack such as client ip addresses, connection handling (features similar to timeout and so on). But imo there are only a few more calls to the gamecontroller missing from the gameserver and maybe a few other places such as entities/ and basing mods on ddnet could be really smooth including updates to new ddnet versions. But sadly that is now off the table.

**Comment by Kaffeine** [comment]
> @ChillerDragon ... custom entities in the world


That's a solved problem and I've linked you a 0.7 solution years ago. Recently I've also pushed 0.6/ddnet (infclass) solution (and also gave you a link somewhere here on ddnet github): https://github.com/infclass/teeworlds-infclassR/blob/6522d8660c30146a75a4cae52f88d85e4137a64e/src/game/server/infclass/ic_entities.cpp#L25. It is almost trivial:
- https://github.com/infclass/teeworlds-infclassR/commit/16720b9b265f50e1e133451d05b31d4678472f54 Use std::vector for entity types
- https://github.com/infclass/teeworlds-infclassR/commit/d7d7b6a4127b3fd1f2e36c75c3add835bd6d9523 Remove all except one NUM_ENTTYPES usages
- https://github.com/infclass/teeworlds-infclassR/commit/ea826856c50f8d09cc2a976e55ca18792dc226e0 Make it possible to register entity types
(and then)
- https://github.com/infclass/teeworlds-infclassR/commit/1885cf66d96930ae59aaa99a4cf046a09e2e96e1

My CGameWorld (as well as CPlayer, CCharacter, et cetera) knows nothing about mod-specific entitites.

The only unnecessarily complicated class is CGameContext.

**Comment by AssassinTee** [comment]
Just to add to this discussion:

Modding based on ddnet is generally a good thing and a lot of fixes/bugs that occur in mods go upstream into ddnet as well, meaning ddnet profits from mods using their codebase (duh!). Multiple things went upstream from unique for example. So my take on this is, to have mods in mind when introducing new features and try not to "put too many stones in their way".

On the other hand sometimes you need to break up things, that mods rely on. For example if mods rely on envelope sync beeing broken. But we should have this in mind and maybe provide a replacement
